map_test.go

```diff
@@ -2,128 +2,513 @@ package swiss
 
 import (
 	"fmt"
-	"math/bits"
 	"math/rand"
-	"strconv"
+	"sort"
 	"testing"
+	"unsafe"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-// TODO(peter)
-// - Insertion
-// - Deletion
-// - Lookup hit vs lookup miss
-// - Resizing
-// - Dropping tombstones without resizing
-// - Iteration
-// - Iteration while mutating the map
+// toBuiltinMap returns the elements as a map[K]V. Useful for testing.
+func (m *Map[K, V]) toBuiltinMap() map[K]V {
+	r := make(map[K]V)
+	m.All(func(k K, v V) bool {
+		r[k] = v
+		return true
+	})
+	return r
+}
 
-func TestBasic(t *testing.T) {
-	m := New[int, int](0)
-	for i := 0; i < 10; i++ {
-		m.Put(i, i+10)
-		if v, ok := m.Get(i); ok {
-			fmt.Printf("%d\n", v)
+// TODO(peter): Extracting a random element might be generally useful. Should
+// this be promoted to the public API?
+func (m *Map[K, V]) randElement() (key K, value V, ok bool) {
+	if m.capacity > 0 {
+		offset := uintptr(rand.Intn(int(m.capacity)))
```

Comment:
For testing it's probably not important but as a public API we probably want equal probability for elements. There will be a few larger gaps within the table and the elements after those would have a higher chance. I would check the group that starts at `i` and if that doesn't work, generate another random number.

Reply:
Good point that we'd want to have a stronger guarantee about what "random element" means in a public API.